src/core/build_env.go

```diff
@@ -220,74 +184,115 @@ func TestEnvironment(state *BuildState, target *BuildTarget, testDir string) Bui
 func ExecEnvironment(state *BuildState, target *BuildTarget, execDir string) BuildEnv {
-	env := TargetEnvironment(state, target)
-
-	env = append(env, "TMP_DIR="+filepath.Join(RepoRoot, execDir))
+	env := append(RuntimeEnvironment(state, target, true, true),
+		"TMP_DIR="+execDir,
+		"TMPDIR="+execDir,
+		"HOME="+execDir,
+		// This is used by programs that use display terminals for correct handling
+		// of input and output in the terminal where the program is run.
+		"TERM="+os.Getenv("TERM"),
+	)
 
 	outEnv := target.Outputs()
+	// OUTS/OUT environment variables being always set is for backwards-compatibility.
+	// Ideally, if the target is a test these variables shouldn't be set.
 	env = append(env, "OUTS="+strings.Join(outEnv, " "))
 	if len(outEnv) == 1 {
-		env = append(env, "OUT="+outEnv[0])
+		env = append(env, "OUT="+resolveOut(outEnv[0], ".", target.Sandbox))
+		if target.IsTest() {
+			env = append(env, "TEST="+resolveOut(outEnv[0], ".", target.Test.Sandbox))
+		}
 	}
 
 	return withUserProvidedEnv(target, env)
 }
 
+// RuntimeEnvironment is the base environment for runtime-based environments.
+// Tools and data env variables are made available.
+func RuntimeEnvironment(state *BuildState, target *BuildTarget, abs, inTmpDir bool) BuildEnv {
+	env := TargetEnvironment(state, target)
+
+	// Data
+	env = append(env, dataEnv(state, target.AllData(), target.NamedData, "", inTmpDir)...)
+
+	if target.IsTest() {
+		// Test tools
+		env = append(env, toolsEnv(state, target.AllTestTools(), target.NamedTestTools(), "", abs)...)
+	}
+
+	if target.Debug != nil {
+		prefix := "DEBUG_"
+		// Debug data
+		env = append(env, dataEnv(state, target.AllDebugData(), target.DebugNamedData(), prefix, inTmpDir)...)
+		// Debug tools
+		env = append(env, toolsEnv(state, target.AllDebugTools(), target.Debug.namedTools, prefix, abs)...)
+	}
+
+	return env
+}
+
+// Handles resolution of OUT files
+func resolveOut(out string, dir string, sandbox bool) string {
+	// Bit of a hack; ideally we would be unaware of the sandbox here.
+	if sandbox && runtime.GOOS == "linux" && !strings.HasPrefix(RepoRoot, "/tmp/") && dir != "." {
+		return path.Join(SandboxDir, out)
+	}
+	return path.Join(dir, out)
+}
+
+// Creates tool-related env variables
+func toolsEnv(state *BuildState, allTools []BuildInput, namedTools map[string][]BuildInput, prefix string, abs bool) BuildEnv {
+	env := BuildEnv{
+		prefix + "TOOLS=" + strings.Join(toolPaths(state, allTools, abs), " "),
+	}
+	if len(allTools) == 1 {
+		env = append(env, prefix+"TOOL="+toolPath(state, allTools[0], abs))
+	}
+	for name, tools := range namedTools {
+		env = append(env, prefix+"TOOLS_"+strings.ToUpper(name)+"="+strings.Join(toolPaths(state, tools, abs), " "))
+	}
+	return env
+}
+
+// Creates data-related env variables
+func dataEnv(state *BuildState, allData []BuildInput, namedData map[string][]BuildInput, prefix string, inTmpDir bool) BuildEnv {
```

Comment:
Refactored data-related env variables into this function. It also fixes the problems where we were previously using the old `target.AllDataPaths(state.Graph)` if `inTmpDir` was true. This function was resolving requires/provides before returning the paths, which was not in line with how we prepare runtime data - where we don't resolve requires/provides.